File: i18n/i18n.py
"""Translation lookup for the Flask apps.

    from i18n import Translator
    t = Translator(catalogue_dir, default="es")
    t("portal.tiles.cameras", locale="fr")
    t("common.minutes_ago", n=5)

English is the fallback. A key with no translation renders the English string
rather than the key name: a bare `portal.tiles.cameras` on a wall panel is worse
than an untranslated word.
"""

from __future__ import annotations

import json
from pathlib import Path


class Translator:
    def __init__(self, catalogue_dir: str | Path, default: str = "en"):
        self.dir = Path(catalogue_dir)
        self.default = default
        self._cache: dict[str, dict[str, str]] = {}
        self._fallback = self._load("en")

    def _load(self, locale: str) -> dict[str, str]:
        if locale in self._cache:
            return self._cache[locale]
        path = self.dir / f"{locale}.json"
        data = json.loads(path.read_text(encoding="utf-8")) if path.exists() else {}
        self._cache[locale] = data
        return data

    @property
    def available(self) -> list[str]:
        """Locales that actually have a catalogue. The admin page offers only
        these, so a half-added language cannot be selected by accident."""
        return sorted(p.stem for p in self.dir.glob("*.json"))

    def name_of(self, locale: str) -> str:
        return self._load(locale).get("_meta.name", locale)

    def __call__(self, key: str, locale: str | None = None, **params) -> str:
        catalogue = self._load(locale or self.default)
        text = catalogue.get(key)
        if text is None:
            text = self._fallback.get(key)
        if text is None:
            text = key
        if params:
            for name, value in params.items():
                text = text.replace("{" + name + "}", str(value))
        return text

    def js(self, key: str, locale: str | None = None, **params) -> str:
        """The same string, safe to drop inside a JavaScript string literal.

        Jinja's autoescaping does not help here. Inside an attribute like
        onclick="...confirm('{{ t(...) }}')", the browser HTML-decodes the
        attribute *before* the JS engine parses it, so an escaped `&#39;` in an
        interpolated name becomes a real quote and closes the literal. Escaping
        for JS first, and letting Jinja escape the result for HTML, is the only
        order that survives both passes.
        """
        text = self(key, locale, **params)
        return (text.replace("\\", "\\\\")
                    .replace("'", "\\'")
                    .replace('"', '\\"')
                    .replace("\n", "\\n")
                    .replace("\r", "")
                    .replace("<", "\\u003c")
                    .replace(">", "\\u003e")
                    .replace("&", "\\u0026"))

File: i18n/i18n.php
<?php
// Translation lookup for the entry page.
//
// Reads the same flat catalogues the Flask apps and the browser dashboards do.
// English is the fallback, for the same reason: an untranslated word beats a
// key name on a screen someone is looking at.

function home_stack_locale_dir(): string {
    return getenv('HOME_STACK_I18N_DIR') ?: __DIR__;
}

function home_stack_catalogue(string $locale): array {
    static $cache = [];
    if (isset($cache[$locale])) return $cache[$locale];
    $path = home_stack_locale_dir() . "/{$locale}.json";
    $cache[$locale] = is_readable($path)
        ? (json_decode(file_get_contents($path), true) ?: [])
        : [];
    return $cache[$locale];
}

function home_stack_available_locales(): array {
    $out = [];
    foreach (glob(home_stack_locale_dir() . '/*.json') as $path) {
        $out[] = basename($path, '.json');
    }
    sort($out);
    return $out;
}

// Order of preference: an explicit ?lang=, then the session, then what the
// browser asks for, then the house default. A panel on a wall never sends
// Accept-Language worth trusting, which is why the configured default wins
// over nothing rather than over everything.
function home_stack_pick_locale(): string {
    $available = home_stack_available_locales();
    $default = getenv('HOME_STACK_DEFAULT_LOCALE') ?: 'en';

    if (isset($_GET['lang']) && in_array($_GET['lang'], $available, true)) {
        if (session_status() === PHP_SESSION_ACTIVE) $_SESSION['lang'] = $_GET['lang'];
        return $_GET['lang'];
    }
    if (isset($_SESSION['lang']) && in_array($_SESSION['lang'], $available, true)) {
        return $_SESSION['lang'];
    }
    foreach (explode(',', $_SERVER['HTTP_ACCEPT_LANGUAGE'] ?? '') as $part) {
        $code = strtolower(substr(trim(explode(';', $part)[0]), 0, 2));
        if (in_array($code, $available, true)) return $code;
    }
    return in_array($default, $available, true) ? $default : 'en';
}

function t(string $key, array $params = [], ?string $locale = null): string {
    $locale = $locale ?: home_stack_pick_locale();
    $text = home_stack_catalogue($locale)[$key]
        ?? home_stack_catalogue('en')[$key]
        ?? $key;
    foreach ($params as $name => $value) {
        $text = str_replace('{' . $name . '}', (string) $value, $text);
    }
    return $text;
}

// Escaped by default. Every call site on a page renders into HTML.
function te(string $key, array $params = [], ?string $locale = null): string {
    return htmlspecialchars(t($key, $params, $locale), ENT_QUOTES, 'UTF-8');
}
